### ISSUE TITLE
TypeError: Class extends value undefined is not a function or null

### ISSUE THREAD

**Issue by prolink007** [typeorm/typeorm#383]
I am getting the following error when trying to create these entities.

`TypeError: Class extends value undefined is not a function or null`

I am assuming this has something to do with circular dependencies, but how is that supposed to be avoided when using table inheritance and one to many relationships?

It is complaining about the following javascript at `BaseComic_1.BaseComic`.

**`let Variant = class Variant extends BaseComic_1.BaseComic {`**

Here is the complete file.

```
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const typeorm_1 = require("typeorm");
const Comic_1 = require("./Comic");
const BaseComic_1 = require("./BaseComic");
let Variant = class Variant extends BaseComic_1.BaseComic {
};
__decorate([
    typeorm_1.ManyToOne(type => Comic_1.Comic, comic => comic.variants),
    __metadata("design:type", Comic_1.Comic)
], Variant.prototype, "comic", void 0);
Variant = __decorate([
    typeorm_1.ClassEntityChild()
], Variant);
exports.Variant = Variant;
//# sourceMappingURL=Variant.js.map
```

```
import {Entity, Column, PrimaryGeneratedColumn, OneToMany} from "typeorm";
import {Comic} from "./Comic";

@Entity()
export class Series {

    @PrimaryGeneratedColumn()
    public id: number;

    @Column("text", {
        length: 30
    })
    public copyright: string;

    @Column("text", {
        length: 100
    })
    public attributionText: string;

    @Column("text", {
        length: 150
    })
    public attributionHTML: string;

    @Column("text", {
        length: 50
    })
    public etag: string;

    @Column("text", {
        length: 200
    })
    public title: string;

    @Column("text")
    public description: string;

    @Column("number", {
        length: 4
    })
    public startYear: number;

    @Column("number", {
        length: 4
    })
    public endYear: number;

    @Column("text", {
        length: 20
    })
    public rating: string;

    @Column("text", {
        length: 20
    })
    public type: string;

    @Column("text")
    public thumbnail: string;

    @OneToMany(type => Comic, comic => comic.series)
    public comics: Array<Comic>;
}
```

```
import {Entity, TableInheritance, PrimaryGeneratedColumn, Column, ManyToOne, DiscriminatorColumn} from "typeorm";
import {Series} from "./Series";

@Entity()
@TableInheritance("class-table")
@DiscriminatorColumn({ name: "type", type: "string"})
export class BaseComic {

    @PrimaryGeneratedColumn()
    public id: number;

    @Column("text", {
        length: 30
    })
    public copyright: string;

    @Column("text", {
        length: 100
    })
    public attributionText: string;

    @Column("text", {
        length: 150
    })
    public attributionHTML: string;

    @Column("text", {
        length: 50
    })
    public etag: string;

    @Column("text", {
        length: 200
    })
    public title: string;

    @Column("int")
    public issue: number;

    @Column("text")
    public variantDescription: string;

    @Column("boolean")
    public variant: boolean;

    @Column("text")
    public description: string;

    @Column("int")
    public pageCount: number;

    @Column("date")
    public onSaleDate: Date;

    @Column("date")
    public unlimitedDate: Date;

    @Column("text")
    public thumbnail: string;

    @ManyToOne(type => Series, series => series.comics)
    public series: Series;
}
```

```
import {OneToMany, ClassEntityChild} from "typeorm";
import {Variant} from "./Variant";
import {BaseComic} from "./BaseComic";

@ClassEntityChild()
export class Comic extends BaseComic {

    @OneToMany(type => Variant, variant => variant.comic)
    public variants: Variant[];
}
```

```
import {ManyToOne, ClassEntityChild} from "typeorm";
import {Comic} from "./Comic";
import {BaseComic} from "./BaseComic";

@ClassEntityChild()
export class Variant extends BaseComic {

    @ManyToOne(type => Comic, comic => comic.variants)
    public comic: Comic;
}
```

**Comment by dotansimha** [comment]
Same here. I guess that the fact that the classes has circular dependency causes the import to return undefined.
Try putting both classes in one file

**Comment by pleerock** [comment]
@dotansimha is correct, you need to change your code structure to prevent this error. Closing.

**Comment by timc13** [comment]
can we get clarification on this?  The [example in the docs](https://github.com/typeorm/typeorm/blob/master/docs/many-to-one-one-to-many-relations.md) has a circular relationship and should encounter this error as well.  Does this mean all related entities have to be declared in the same file?  This seems suboptimal.  We should be able to declare relations using strings to avoid this circular import.

**Comment by Prozi** [comment]
found out that i had

```
(file1.js)
module.exports.default = class XXX {
...
}

(file2.js)
const XXX = require('./file1')

class YYY extends XXX {
...
}
```

ass you see I forgot to add `.default` in `require`

it wasn't circular at all it just wasnt having the main class since `require` yielded `undefined`

**Comment by Alfagun74** [comment]
i just had to delete my dist folder once to resolve this issue.

**Comment by spedroq** [comment]
I was running into this issue myself, and this is how I overcame it. 
__TLDR:__ What I did was ensure that the relations are declared in the correct place.

## Example
For example, if you have the following entities:
 - User
 - Address
 - Contact
 - Item
 
 And the following relations:
  - User has locations and contacts
  - Address was created/updated by user
  - Contact was created/updated by user
  - Item was created/updated by user
  
 ## Solution
  What you need to make sure is that the user entity itself does not mention the others. Instead to create the relation between them, place that ownership on the other entity.
  
 ### Circular Dependency Issue Example
  For example, initially I had something similar to this:
```js
@Entity("user")
export class UserEntity {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column("text")
  name: string;

  @Column("text", { unique: true })
  email: string;

  @ManyToMany(() => ContactEntity, (contact) => contact.id)
  @JoinTable({ name: "user_contact" })
  @Optional()
  contacts?: ContactEntity[];
}
  ```
  and the contact entity:
  ```js
    @Entity("contact")
    export class ContactEntity extends BaseEntity {
    
      @Column("text")
      contact: string;
 
      @ManyToOne(() => UserEntity, (user) => user.id)
      @JoinColumn({ name: "createdBy" })
      createdBy: UserEntity;
    
      @ManyToOne(() => UserEntity, (user) => user.id, { nullable: true })
      @JoinColumn({ name: "updatedBy" })
      updatedBy?: UserEntity;
    }
  ```
  ### Solved Example
  Then fixed it by moving the relation to the contact instead of the user, since user is used everywhere else, this avoids the circular dependency:
  ```js
@Entity("user")
export class UserEntity {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column("text")
  name: string;

  @Column("text", { unique: true })
  email: string;

  @ManyToMany(() => ContactEntity, (contact) => contact.id)
  @JoinTable({ name: "user_contact" })
  @Optional()
  contacts?: ContactEntity[];
}
  ```
  and the contact entity:
  ```js
  @Entity("contact")
  export class ContactEntity extends BaseEntity {
  
    @Column("text")
    contact: string;
    
    @ManyToOne(() => UserEntity, (user) => user.id)
    @JoinColumn({ name: "createdBy" })
    createdBy: UserEntity;
  
    @ManyToOne(() => UserEntity, (user) => user.id, { nullable: true })
    @JoinColumn({ name: "updatedBy" })
    updatedBy?: UserEntity;
  
    @ManyToMany(() => UserEntity, (user) => user.id, { nullable: true })
    @JoinTable({ name: "user_contact" })
    users?: UserEntity[];
  }
  ```
  
 ## Conclusion
  Hope this helps anyone getting this error. I was running into this issue since I had made a base class for the auditing properties of createdBy/updatedBy etc. And this was causing me circular dependencies since everything used it, and since my user had relations to entities that had already extended it, it was causing that circular import issue.

**Comment by swaaap-nil** [comment]
I was also facing this issue and wasted so much time. I am using nestjs. For me the soln was to tell the compiler to only pickup either .ts or .js files. ===>  entities: [path.join(__dirname, '../**/*.entity.ts')], if i also include .js files it reads code for same entity twice and gets confused.

**Comment by alumni** [comment]
@spedroq the solution provided above is still not correct, the second parameter of e.g. `ManyToOne` is the inverse relation, not the primary key of the references entity. You can also omit it if there is no inverse.
